### DISCUSSION TITLE
keyboard.send_key("<tab>") stopped working!

### DISCUSSION THREAD

**Discussion by pelliott80** [autokey/autokey#547] [Q&A]
keyboard.send_key("<tab>") stopped working.

In 2019 I wrote a script that took the contents of the clipboard, split it into TAB delimited strings, and then sent the data to the keyboard, Each field being delimited by keyboard.send_key("<tab>"). The keyboard.send_key("<tab") would cause a browser to switch fields exactly as if TAB had been typed in the keyboard. (Control-v does not do that. Control-v interprets TAB characters as data entered directly into the field.) That code was debugged and it worked in 2019. But there has been an new release of Debian since then, and the code stopped working. Under KDE plasma the TABS are turning into spaces, that do not affect the browser behavior. Under Gnome, the TABS are being ignored as if they were not there! I have tried this code also under Ubuntu where it also does not work.

Can anyone suggest what is going wrong? Perhaps the TABS are being intercepted by the window manager? Can anyone suggest a work-around? What am I doing wrong? This code used to work. If someone can not tell me what I am doing wrong, I will re-submit this as an issue. Thank You for looking at this.

I include my code below:

```
import time


# We have not done anything yet, so 
# NotFirstDone is false.
FirstDone = False

#split the clipboard contents into a
#list of TAB delimited strings
#For each string.....
for entry in  re.split('\t',clipboard.get_clipboard()):
    # if we have done a string before....
    #First entry we will skip sending a TAB
    if FirstDone:
        # send a TAB to move to next field
        keyboard.send_key("<tab>")

    #sleep to avoid problems    
    time.sleep(0.25)
    #send the current string as if it came from keyboard
    keyboard.send_keys(entry)
    #sleep to avoid problems
    time.sleep(0.25)
    #We have now Done at least 1 string.
    #next loop pass (and all others), 
    #will send a tab first
    Firstdone=True

```

**Comment by josephj11** [comment]
(I tried to cover everything in this answer. That's why it's so long. Hopefully, it will end up in our Wiki.)

TL;DR: See if using `keyboard.send_keys()` fixes it. :) If not, see below.

The AutoKey API calls communicate directly with the active window, so, usually, the desktop environment (DTE) will not see these keypresses. This is usually a good thing, but it means that normally, you can not intentionally talk to the DTE.

When in doubt, and any time that prolonged staring at your code doesn't reveal why it is not doing what you want it to, running an AutoKey internal [trace](https://github.com/autokey/autokey-python2/wiki/Problem-Reporting-Guide) will show you things from AutoKey's perspective. Traces are also usually valuable additions to new issue reports. Sometimes they are essential for us to understand and recreate an issue.

Let's take this a step at a time. 
* When you post on GitHub or on [Gitter](https://gitter.im/autokey/autokey), your text is processed as markdown text, so if anything can be misinterpreted as markdown codes, it will be. I extracted your code, undenteded it, and surrounded it with lines that contain only triple backtics. This makes it a block quote that will not be processed for markdown. GitHub will also do that for you if select the text and then select the code button with the `<>` icon. Anyway, now your code displays correctly.

When emitting text from a script try these API calls in this order and use the first method that meets your needs:
*   `clipboard.fill_clipboard("your text")`
    `keyboard.send_keys("<ctrl>+v")`

This is the most bulletproof method, but (as in your case) it will not process active keys like tabs.
This method
 *  Allows the sending of arbitrarily long texts
  * Avoids a lot of race/timing issues with slow applications
  * Usually handles multibyte characters such as non-EN-US character sets correctly

If unexpected results occur when using the clipboard API, the most frequent cause is that both the Qt and GTK clipboard libraries that AutoKey uses return success instantly while the underlying processes are still working.
Adding small delays before and/or after the clipboard API calls usually resolves these issues.  `time.sleep(floating point delay in seconds)` does this.

* `keyboard.send_keys("your text")`

This will process active keys, but will not work for multibyte characters such as almost anything that is not EN-US or *almost* anything that you cannot directly type on your physical keyboard. (Note the `s` on the end of the API call function name.)
Sometimes, it does have problems with long texts. How long "long" is is largely dependent on the receiving application.
When sending active keys such as tabs, it often helps to insert delays between events that cause actions and subsequent character emissions. `time.sleep(floating point delay in seconds)` does this.

* `keyboard.send_key("your single character")`

Only use this when nothing else works such as when you need its optional repeat parameter.
It *really* means a single key. It won't even emit upper case or shifted/modified characters.

If things still don't work as desired, then you have to go down the rabbit hole by calling an external tool such as `xdotool` or `Tkinter`, coding in Python or directly accessing libraries such as those provided by GTK or Qt.

For more information on emitting Unicode characters, see [this](https://github.com/autokey/autokey/wiki/Adding-Unicode-Characters-or-Emojis-to-Your-Scripts-or-Phrases) wiki page.

For ways to do some of these things and other advanced actions with AutoKey phrases, see [this](https://github.com/autokey/autokey/wiki/Dynamic-Phrases,-Using-Macros-as-placeholders-in-Phrases) wiki page.
